### PR TITLE
Consolidate binary and logical functions

### PR DESCRIPTION
**Pull request by arnikola**
Consolidate binary and logical functions into a single package. Reduce reused code and generalize creation of binary functions, making creation of any future binary functions easier.

### CODE REVIEW

File: src/query/functions/binary/binary.go [hunk 1]
```diff
@@ -127,98 +142,47 @@ var (
...
+		return processBinary(lhs, rhs, params, controller, fn)
 	}
-
-	defer nextBlock.Close()
-	return n.controller.Process(nextBlock)
 }
 
 // processes two logical blocks, performing a logical operation on them
-func (n *binaryNode) process(lhs, rhs block.Block) (block.Block, error) {
+func processBinary(
```
Comment (nikunjgit) [inline src/query/functions/binary/binary.go]: the signature could become simpler if this was defined on the node object as before.

File: src/query/functions/binary/binary.go [hunk 2]
```diff
@@ -113,12 +114,26 @@ var (
...
+		ExpType + returnBoolSuffix:      math.Pow,
+
+		EqType:        func(x, y float64) float64 { return toComparisonValue(x == y, x) },
+		NotEqType:     func(x, y float64) float64 { return toComparisonValue(x != y, x) },
+		GreaterType:   func(x, y float64) float64 { return toComparisonValue(x > y, x) },
+		LesserType:    func(x, y float64) float64 { return toComparisonValue(x < y, x) },
+		GreaterEqType: func(x, y float64) float64 { return toComparisonValue(x >= y, x) },
+		LesserEqType:  func(x, y float64) float64 { return toComparisonValue(x <= y, x) },
+
+		EqType + returnBoolSuffix:        func(x, y float64) float64 { return toFloat(x == y) },
```
[the remark this answers is not in the document]
Comment (arnikola) [reply src/query/functions/binary/binary.go]: That's how I had it originally, but I'd prefer to incur a small cost here in adding unnecessary types to the map than go through an additional function call for every value in the series

File: src/query/functions/binary/binary.go [hunk 3]
```diff
@@ -127,98 +142,47 @@ var (
...
-
-	nextBlock, err := n.process(lhs, rhs)
-	if err != nil {
-		return err
+// Build the binary processing step
+func createBinaryProcessingStep(
+	params NodeParams,
+	fn mathFunc,
+) func(block.Block, block.Block, *transform.Controller) (block.Block, error) {
+	return func(lhs, rhs block.Block, controller *transform.Controller) (block.Block, error) {
```
[the remark this answers is not in the document]
Comment (arnikola) [reply src/query/functions/binary/binary.go]: This way I can have a more generic base node which can accept both logical and binary functions

File: src/query/functions/binary/binary.go (hunk 1 above)
Comment (arnikola) [reply src/query/functions/binary/binary.go]: I'd prefer a more complex function generator signature in exchange for a simpler node

File: src/query/functions/binary/binary.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nikunjgit) [reply src/query/functions/binary/binary.go]: problem now is that its easy to make a mistake by adding the type but not adding type + returnBoolSuffix. Also, you're writing the same logic twice

[the remark this answers is not in the document]
Comment (nikunjgit) [reply src/query/functions/binary/binary.go]: also you can't really avoid additional call. Even here you're calling toComparisonValue

File: src/query/functions/binary/binary.go (hunk 1 above)
Comment (nikunjgit) [reply src/query/functions/binary/binary.go]: but then you're passing the same arguments multiple layers deep. This provides an interesting comparison: https://stackoverflow.com/questions/36904955/passing-single-object-vs-passing-multiple-parameters

File: src/query/functions/binary/binary.go [hunk 4]
```diff
@@ -347,49 +275,34 @@ func (n *binaryNode) processSingleBlock(
...
-		}
-		// NB(arnikola): whereas prometheus does not include not matching
-		// points in the output, we here replace them with NaNs instead
-		return math.NaN()
-	}
-}
-
-func (n *binaryNode) processBothSeries(
+func processBothSeries(
 	lIter, rIter block.StepIter,
```
Comment (nikunjgit) [inline src/query/functions/binary/binary.go]: this functions has 6 arguments now. Consider object ? https://softwareengineering.stackexchange.com/questions/145055/are-there-guidelines-on-how-many-parameters-a-function-should-accept

File: src/query/functions/binary/binary.go [hunk 5]
```diff
@@ -22,14 +22,10 @@ package binary
 
 import (
```
Comment (nikunjgit) [inline src/query/functions/binary/binary.go]: binary/binary.go is a bit strange. Consider renaming ?

File: src/query/functions/binary/base.go [hunk 6]
```diff
@@ -29,76 +29,76 @@ import (
...
-	case UnlessType:
-		makeBlock = makeUnlessBlock
-	default:
-		return logicalOp{}, fmt.Errorf("operator not supported: %s", opType)
+	fn, ok := buildLogicalFunction(opType, params)
+	if !ok {
+		fn, ok = buildArithmeticFunction(opType, params)
+		if !ok {
+			fn, ok = buildComparisonFunction(opType, params)
+			if !ok {
```
Comment (benraskin92) [inline src/query/functions/binary/base.go]: Anyway to clean this code up a bit? Looks odd having 3 nested `if` statements.

File: src/query/functions/binary/binary.go [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+	controller *transform.Controller,
+	fn binaryFunc,
+) (block.Block, error) {
+	lIter, err := lhs.StepIter()
+	if err != nil {
+		return nil, err
+	}
+
+	// fn := n.op.fn
+	// params := n.op.params
```
Comment (benraskin92) [inline src/query/functions/binary/binary.go]: Remove?

File: src/query/functions/binary/base.go (hunk 6 above)
Comment (arnikola) [reply src/query/functions/binary/base.go]: I tried a few things but unfortunately there's not many ways to add sugar to functions that return 2 values. Best alternative I came up with was something like 
```
fn, ok = buildArithmeticFunction(opType, params)
		if !ok {
			fn, ok = buildBinaryFunction(opType, params)
                        if !ok {
                          return error
```
and have buildBinaryFunction try building an Arithmetic then a Comparison function, but that doesn't achieve much except for kicking the can down the road

File: src/query/functions/binary/binary.go (hunk 7 above)
Comment (arnikola) [reply src/query/functions/binary/binary.go]: Oops, missed this. Will do it in a future commit

**Pull request merged**